redwood-lazylayout-uiview/src/commonMain/kotlin/app/cash/redwood/lazylayout/uiview/UIViewLazyList.kt

```diff
@@ -26,101 +26,197 @@ import app.cash.redwood.layout.api.Constraint
+import platform.UIKit.UICollectionViewCell
+import platform.UIKit.UICollectionViewDataSourceProtocol
+import platform.UIKit.UICollectionViewDelegateFlowLayoutProtocol
+import platform.UIKit.UICollectionViewFlowLayout
+import platform.UIKit.UICollectionViewLayout
+import platform.UIKit.UICollectionViewScrollDirection
 import platform.UIKit.UIControlEventValueChanged
+import platform.UIKit.UIEdgeInsetsMake
+import platform.UIKit.UILabel
 import platform.UIKit.UIRefreshControl
-import platform.UIKit.UITableView
-import platform.UIKit.UITableViewCell
-import platform.UIKit.UITableViewCellStyle
-import platform.UIKit.UITableViewCellStyle.UITableViewCellStyleDefault
-import platform.UIKit.UITableViewDataSourceProtocol
-import platform.UIKit.UITableViewDelegateProtocol
+import platform.UIKit.UIScrollView
 import platform.UIKit.UIView
+import platform.UIKit.indexPathForRow
 import platform.UIKit.item
 import platform.darwin.NSInteger
 import platform.darwin.NSObject
 
-private const val reuseIdentifier = "cell"
+internal interface ViewPortItems: Widget.Children<UIView> {
+  var placeholders: MutableList<Widget<UIView>>
+  var itemsBefore: Int
+  var itemsAfter: Int
+  fun itemForGlobalIndex(index: Int): UIView
+  fun itemCount(): Int
+}
+
+internal open class UIViewLazyList() : LazyList<UIView>, ChangeListener {
 
-internal open class UIViewLazyList() : LazyList<UIView> {
+  override val items: ViewPortItems = object : ViewPortItems {
+    override var placeholders = mutableListOf<Widget<UIView>>()
 
-  private val itemsList = mutableListOf<Widget<UIView>>()
+    override var itemsBefore: Int = 0
+    override var itemsAfter: Int = 0
 
-  override val placeholder: Widget.Children<UIView> = MutableListChildren()
+    private val viewPortList = mutableListOf<Widget<UIView>>()
 
-  override val items: Widget.Children<UIView> = object : Widget.Children<UIView> {
     override fun insert(index: Int, widget: Widget<UIView>) {
-      itemsList.add(index, widget)
-      tableView.reloadData()
+      viewPortList.add(index, widget)
     }
 
     override fun move(fromIndex: Int, toIndex: Int, count: Int) {
-      itemsList.move(fromIndex, toIndex, count)
-      tableView.reloadData()
+      viewPortList.move(fromIndex, toIndex, count)
     }
 
     override fun remove(index: Int, count: Int) {
-      itemsList.remove(index, count)
-      tableView.reloadData()
+      viewPortList.remove(index, count = count)
+    }
+
+    override fun onModifierUpdated() {}
+
+    // Fetch the item from the viewPortList relative to the entire collection view
+    override fun itemForGlobalIndex(index: Int): UIView {
+      val viewPortIndex: Int = max(index - itemsBefore, 0)
+
+      viewPortList.getOrNull(viewPortIndex)?.let {
+        return it.value
+      }
+
+      // If we don't have a value, fallback to our pools of placeholders
+      val placeholderIndex = index % placeholders.size
+      return placeholders.get(placeholderIndex).value
```

Comment:
When the `placeholders.size` is less than the number of items that can fill the view port, we might be reusing an already visible placeholder. Is that a big deal on iOS?

Reply:
it's not ideal, but i think it beats maintaining a pool for a really niche case where we have many small cells. we could use the viewport size to increase the initial pool size if needed